### PR TITLE
reverse binding order in matches to allow the subbinding of copyable fields in bindings after @

### LINKED ISSUE #69971: Unnecessary 'cannot bind by-move with sub-bindings' with `bindings_after_at`
The following code [(playground link)](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=fd86d6194ad9c73b31e4e9032e782d62) attempts to move a struct while using a sub-binding to copy one of its fields:

```rust
#![feature(bindings_after_at)]

struct NonCopyStruct {
    copy_field: u32,
}

fn foo(x: NonCopyStruct) {
    let y @ NonCopyStruct { copy_field: z } = x;
}
```

On nightly, it produces two errors:

```
error[E0007]: cannot bind by-move with sub-bindings
 --> src/lib.rs:8:9
  |
8 |     let y @ NonCopyStruct { copy_field: z } = x;
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ binds an already bound by-move value by moving it
error[E0382]: use of moved value: `x`
 --> src/lib.rs:8:41
  |
7 | fn foo(x: NonCopyStruct) {
  |        - move occurs because `x` has type `NonCopyStruct`, which does not implement the `Copy` trait
8 |     let y @ NonCopyStruct { copy_field: z } = x;
  |         --------------------------------^--
  |         |                               |
  |         |                               value used here after move
```

However, it would be sound to allow this, since it's equivalent to copying the field before or after moving the struct.

(There are no errors if the struct does implement `Copy`.)

### PR DESCRIPTION
**Pull request by vn-ki**
Fixes #69971

### TODO

- [x] Regression tests

r? @oli-obk

### CODE REVIEW

File: src/test/ui/pattern/bindings-after-at/borrowck-pat-at-and-box.rs [hunk 1]
```diff
@@ -18,53 +18,51 @@ fn nc() -> NC {
...
-    //~^ ERROR use of moved value
 
     fn f1(a @ box &b: Box<&C>) {}
-    //~^ ERROR use of moved value
 
     fn f2(a @ box b: Box<C>) {}
-    //~^ ERROR use of moved value
 
     match Box::new(C) {
-        a @ box b => {} //~ ERROR use of moved value
```
Comment (vn-ki) [inline src/test/ui/pattern/bindings-after-at/borrowck-pat-at-and-box.rs:33]: These patterns are all of a sudden legal now.

File: src/test/ui/pattern/move-ref-patterns/move-ref-patterns-dynamic-semantics.rs [hunk 2]
```diff
@@ -74,6 +74,6 @@ fn main() {
         };
         lam((mk(19), mk(20), mk(21), mk(22)));
     }
-    let expected = [2, 3, 6, 5, 7, 8, 12, 11, 9, 10, 18, 13, 14, 15, 16, 17, 21, 19, 20, 22, 4, 1];
+    let expected = [2, 3, 6, 5, 7, 8, 12, 11, 9, 10, 13, 18, 14, 15, 16, 17, 19, 21, 20, 22, 4, 1];
```
Comment (vn-ki) [inline src/test/ui/pattern/move-ref-patterns/move-ref-patterns-dynamic-semantics.rs:77]: The drop order seems to have changed. I don't know enough about the semantics to know if this is valid.

File: src/test/ui/pattern/bindings-after-at/borrowck-pat-at-and-box.rs (hunk 1 above)
Comment (oli-obk) [reply src/test/ui/pattern/bindings-after-at/borrowck-pat-at-and-box.rs:33]: These are expected. The inner thing is now first copied and then the outer thing is moved. `C` is `Copy`, so this works out.

File: src/test/ui/pattern/bindings-after-at/copy-and-move-mixed.rs [hunk 3]
```diff
@@ -8,10 +8,6 @@ struct C;
 struct NC<A, B>(A, B);
 
 fn main() {
-    let a @ NC(b, c) = NC(C, C);
```
Comment (oli-obk) [inline src/test/ui/pattern/bindings-after-at/copy-and-move-mixed.rs:11]: you lost this test

File: src/test/mir-opt/early_otherwise_branch.opt1.EarlyOtherwiseBranch.diff [hunk 4]
```diff
@@ -52,13 +52,13 @@
 -     }
 - 
 -     bb3: {
-          StorageLive(_8);                 // scope 0 at $DIR/early_otherwise_branch.rs:5:15: 5:16
-          _8 = (((_3.0: std::option::Option<u32>) as Some).0: u32); // scope 0 at $DIR/early_otherwise_branch.rs:5:15: 5:16
```
Comment (oli-obk) [inline src/test/mir-opt/early_otherwise_branch.opt1.EarlyOtherwiseBranch.diff:56]: These tuple patterns swapped their order, that seems wrong (wrt drop order and such), though I can't tell how it's happening, did you bless the mir-opt tests?

File: src/test/ui/pattern/bindings-after-at/copy-and-move-mixed.rs (hunk 3 above)
Comment (vn-ki) [reply src/test/ui/pattern/bindings-after-at/copy-and-move-mixed.rs:11]: it's a passing test. I wanted to add it pass. (part of TODO)

File: src/test/mir-opt/early_otherwise_branch.opt1.EarlyOtherwiseBranch.diff (hunk 4 above)
Comment (vn-ki) [reply src/test/mir-opt/early_otherwise_branch.opt1.EarlyOtherwiseBranch.diff:56]: yup. i blessed mir-opt.

File: compiler/rustc_mir_build/src/build/matches/simplify.rs [hunk 5]
```diff
@@ -44,12 +44,36 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         candidate: &mut Candidate<'pat, 'tcx>,
     ) -> bool {
         // repeatedly simplify match pairs until fixed point is reached
+        debug!("simplify_candidate(candidate={:?})", candidate);
+
+        // exisiting_bindings and new_bindings exists to keep the semantics in order
+        // reversing the binding order for bindings after `@` change binding order in places
+        // it shouldn't be changed, for example `let (Some(a), Some(b)) = (x, y)`
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/build/matches/simplify.rs:51]:
```suggestion
        // exisiting_bindings and new_bindings exists to keep the semantics in order.
        // Reversing the binding order for bindings after `@` changes the binding order in places
        // it shouldn't be changed, for example `let (Some(a), Some(b)) = (x, y)`
```

File: compiler/rustc_mir_build/src/build/matches/simplify.rs [hunk 6]
```diff
@@ -44,12 +44,36 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
+        // * after all iterations, these new bindings are then appended to the bindings that were
+        // prexisting (i.e. `candidate.binding` when the function was called).
+        //
+        // example:
+        // candidate.bindings = [1, 2, 3]
+        // binding in iter 1: [4, 5]
+        // binding in iter 2: [6, 7]
+        //
+        // final binding: [1, 2, 3, 6, 7, 4, 5]
+        let mut exisiting_bindings = mem::take(&mut candidate.bindings);
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/build/matches/simplify.rs:67]:
```suggestion
        let mut existing_bindings = mem::take(&mut candidate.bindings);
```

File: compiler/rustc_mir_build/src/build/matches/simplify.rs [hunk 7]
```diff
@@ -65,13 +89,33 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                     }
                 }
             }
+            // issue #69971: the binding order should be right to left if there are more
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/build/matches/simplify.rs:92]:
```suggestion
            // Avoid issue #69971: the binding order should be right to left if there are more
```

Comment (oli-obk) [reply compiler/rustc_mir_build/src/build/matches/simplify.rs:92]: Before it sounded like this is the reason the issue stays open

File: compiler/rustc_mir_build/src/build/matches/simplify.rs [hunk 8]
```diff
@@ -65,13 +89,33 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
             if !changed {
+                exisiting_bindings.extend_from_slice(&new_bindings);
+                mem::swap(&mut candidate.bindings, &mut exisiting_bindings);
                 // Move or-patterns to the end, because they can result in us
                 // creating additional candidates, so we want to test them as
                 // late as possible.
                 candidate
                     .match_pairs
                     .sort_by_key(|pair| matches!(*pair.pattern.kind, PatKind::Or { .. }));
+                debug!("simplify_candidate: simplifed {:?}", candidate);
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/build/matches/simplify.rs:118]:
```suggestion
                debug!(simplified = ?candidate, "simplify_candidate");
```

File: compiler/rustc_mir_build/src/build/matches/simplify.rs [hunk 9]
```diff
@@ -44,12 +44,36 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         candidate: &mut Candidate<'pat, 'tcx>,
     ) -> bool {
         // repeatedly simplify match pairs until fixed point is reached
+        debug!("simplify_candidate(candidate={:?})", candidate);
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/build/matches/simplify.rs:47]:
```suggestion
        debug!(?candidate, "simplify_candidate");
```

or put

```
#[tracing::instrument(skip(self))]
```

onto the function

File: src/test/ui/pattern/bindings-after-at/copy-and-move-mixed.rs (hunk 3 above)
Comment (oli-obk) [reply src/test/ui/pattern/bindings-after-at/copy-and-move-mixed.rs:11]: still, it doesn't hurt to keep it here to show that this passes in contrast to the non-passing version below

**Review by oli-obk — CHANGES_REQUESTED** [review]

**Comment by oli-obk** [comment]
cc @matthewjasper

**Comment by oli-obk** [comment]
I can't tell why that test would fail only on macs. And especially how the changes in this PR could cause them

@bors retry

**Comment by oli-obk** [comment]
```
Warning: Failed to download action 'https://api.github.com/repos/actions/checkout/zipball/50fbc622fc4ef5163becd7fab6573eac35f8462e'. Error: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.
Warning: Back off 17.33 seconds before retry.
Warning: Failed to download action 'https://api.github.com/repos/actions/checkout/zipball/50fbc622fc4ef5163becd7fab6573eac35f8462e'. Error: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.
Warning: Back off 11.692 seconds before retry.
Error: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.
```

@bors retry

**Pull request merged**